File: src/main/java/at/apf/reallystrangechess/mapper/BoardConverter.java
package at.apf.reallystrangechess.mapper;

import at.apf.reallystrangechess.model.*;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.stream.Stream;

@Component
public class BoardConverter {


    public List<BoardField> fromString(String str) {
        List<BoardField> board = new ArrayList<>();

        String[] lines = str.split("\n");
        int i = 0;

        // board color fields
        while (i < lines.length) {
            String line = lines[i].toLowerCase();
            if (line.isEmpty()) {
                i++;
                break;
            }
            for (int j = 0; j < line.length(); j++) {
                BoardFieldColor color = BoardFieldColor.EMPTY;
                if (line.charAt(j) == 'w') {
                    color = BoardFieldColor.WHITE;
                } else if (line.charAt(j) == 'b') {
                    color = BoardFieldColor.BLACK;
                }
                board.add(new BoardField(new Position(j + 1, i + 1), color, null, null));
            }
            i++;
        }

        // figures
        int i2 = 0;
        while (i < lines.length) {
            String line = lines[i];
            if (line.isEmpty()) {
                i++;
                break;
            }
            for (int j = 0; j < line.length(); j++) {
                Figure fig = null;
                final String c = (line.charAt(j) + "").toUpperCase();
                FigureType type = Stream.of(FigureType.values()).filter(t -> t.getChar().equals(c)).findFirst().orElse(null);
                if (type != null) {
                    Color color = c.equals(line.charAt(j) + "") ? Color.BLACK : Color.WHITE;
                    fig = new Figure(color, type);
                }
                Position pos = new Position(j + 1, i2 + 1);
                Optional<BoardField> field = board.stream().filter(f -> f.getPosition().equals(pos)).findFirst();
                if (field.isPresent() && fig != null) {
                    field.get().setFigure(fig);
                }
            }
            i++;
            i2++;
        }

        return board;
    }
}

File: src/main/webapp/src/util/BoardConverter.ts
import {BoardField, BoardFieldColor, Figure, FigureType, Position} from "../dto/dtos";
import {posEq} from "../logic/BaseChessLogic";


export default function boardFromString(str: string): Array<BoardField> {
    const board: Array<BoardField> = []
    const lines = str.split('\n')

    let i = 0

    // Field colors
    while (i < lines.length) {
        const line = lines[i].toLocaleLowerCase()
        if (line.length === 0) {
            i++
            break;
        }
        for (let j = 0; j < line.length; j++) {
            let color: BoardFieldColor = 'EMPTY'
            if (line.charAt(j) === 'w') {
                color = 'WHITE'
            } else if (line.charAt(j) === 'b') {
                color = 'BLACK'
            }
            board.push({position: {x: j + 1, y: i + 1}, color: color, figure: null, mine: null})
        }
        i++
    }

    // white figures
    let i2 = 0
    while (i < lines.length) {
        const line = lines[i].toLowerCase()
        if (line.length === 0) {
            i++
            break;
        }
        for (let j = 0; j < line.length; j++) {
            let fig: Figure | null = null
            const c = line.charAt(j)
            let type: FigureType | null = null
            if (c === 'b') {
                type = "BAUER"
            } else if (c === 'l') {
                type = "LAUFER"
            } else if (c === 's') {
                type = "SPRINGER"
            } else if (c === 't') {
                type = "TURM"
            } else if (c === 'd') {
                type = "DAME"
            } else if (c === 'k') {
                type = "KING"
            }
            if (type != null) {
                fig = {color: "WHITE", type: type}
            }
            const pos: Position = {x: j + 1, y: i2 + 1}
            const field: BoardField | undefined = board.find(f => posEq(f.position, pos))
            if (field !== undefined && fig !== null) {
                field.figure = fig
            }
        }
        i++
        i2++
    }

    // black figures
    i2 = 0
    while (i < lines.length) {
        const line = lines[i].toLowerCase()
        if (line.length === 0) {
            i++
            break;
        }
        for (let j = 0; j < line.length; j++) {
            let fig: Figure | null = null
            const c = line.charAt(j)
            let type: FigureType | null = null
            if (c === 'b') {
                type = "BAUER"
            } else if (c === 'l') {
                type = "LAUFER"
            } else if (c === 's') {
                type = "SPRINGER"
            } else if (c === 't') {
                type = "TURM"
            } else if (c === 'd') {
                type = "DAME"
            } else if (c === 'k') {
                type = "KING"
            }
            if (type != null) {
                fig = {color: "BLACK", type: type}
            }
            const pos: Position = {x: j + 1, y: i2 + 1}
            const field: BoardField | undefined = board.find(f => posEq(f.position, pos))
            if (field !== undefined && fig !== null) {
                field.figure = fig
            }
        }
        i++
        i2++
    }

    return board
}